src/data/bucket/pattern_attributes.js

```diff
@@ -4,5 +4,7 @@ import {createLayout} from '../../util/struct_array';
 export default createLayout([
     // [tl.x, tl.y, br.x, br.y]
     {name: 'a_pattern_from', components: 4, type: 'Uint16'},
-    {name: 'a_pattern_to', components: 4, type: 'Uint16'}
+    {name: 'a_pattern_to', components: 4, type: 'Uint16'},
+    {name: 'a_pixel_ratio_from', components: 1, type: 'Uint8'},
+    {name: 'a_pixel_ratio_to', components: 1, type: 'Uint8'},
```

Comment:
Can these two new attributes be combined into a single attribute? each attribute is aligned to a 4-byte boundary so putting them into one will save 4 bytes per vertex

Or maybe even packed into the existing attributes? I think it's probably *not* worth it at this point but it might be worth keeping in mind if we ever need to optimize this further.

Reply:
That's a good point! I was thinking about this but assumed it was not the common case. In my testing we didn't hit this code path in the default styles but when using data driven styling in expressions such as `"line-pattern": ["case", ["==", ["get", "class"], "street"], "generic_icon", "generic_metro" ]`.

I'm ok to give that an afterthought and profile more to see if the frequency at which we hit that justify this optimization.

Reply:
Captured it for follow up in https://github.com/mapbox/mapbox-gl-js/issues/9407.